File: frontend/app/src/main/java/com/travelify/auth/viewmodel/AuthViewModel.kt
package com.travelify.auth.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.travelify.auth.api.ApiClient
import com.travelify.auth.api.AuthService
import com.travelify.auth.api.LoginRequest
import com.travelify.auth.util.TokenManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class AuthViewModel(application: Application) : AndroidViewModel(application) {

    private val tokenManager = TokenManager(application)
    private val authService: AuthService = ApiClient.create(tokenManager)

    private val _isLoggedIn = MutableStateFlow(false)
    val isLoggedIn: StateFlow<Boolean> = _isLoggedIn

    init {
        checkForToken()
    }

    private fun checkForToken() {
        viewModelScope.launch {
            _isLoggedIn.value = tokenManager.getToken() != null
        }
    }

    fun login(email: String, pass: String) {
        viewModelScope.launch {
            try {
                val response = authService.login(LoginRequest(email, pass))
                tokenManager.saveToken(response.token)
                _isLoggedIn.value = true
            } catch (e: Exception) {
                // Handle login error
                _isLoggedIn.value = false
            }
        }
    }

    fun logout() {
        viewModelScope.launch {
            tokenManager.deleteToken()
            _isLoggedIn.value = false
        }
    }
}

File: frontend/ios/Travelify/Auth/AuthViewModel.swift
import Foundation
import Combine

class AuthViewModel: ObservableObject {
    @Published var isLoggedIn = false
    private let authService = AuthService()
    private let keychainHelper = KeychainHelper.shared

    init() {
        checkForToken()
    }

    func checkForToken() {
        if keychainHelper.getToken() != nil {
            DispatchQueue.main.async {
                self.isLoggedIn = true
            }
        }
    }

    func login(email: String, pass: String) {
        let loginRequest = LoginRequest(email: email, pass: pass)
        authService.login(loginRequest: loginRequest) { result in
            switch result {
            case .success(let response):
                self.keychainHelper.saveToken(response.token)
                DispatchQueue.main.async {
                    self.isLoggedIn = true
                }
            case .failure(let error):
                // Handle login error
                print(error.localizedDescription)
                DispatchQueue.main.async {
                    self.isLoggedIn = false
                }
            }
        }
    }

    func logout() {
        keychainHelper.deleteToken()
        DispatchQueue.main.async {
            self.isLoggedIn = false
        }
    }
}